File: Camera2Test/app/src/main/java/com/hank/camera2test/utils/PermissionUtils.java
package com.hank.camera2test.utils;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Build;
import android.util.Log;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

public class PermissionUtils {
    private static final String TAG = PermissionUtils.class.getSimpleName();

    public static final String[] sRequiredPermissions = {
            Manifest.permission.CAMERA,
            Manifest.permission.RECORD_AUDIO,
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.ACCESS_FINE_LOCATION
    };

    private static PermissionUtils sInstance;

    public static PermissionUtils getInstance(Context context) {
        if (sInstance == null) {
            sInstance = new PermissionUtils(context);
        }
        return sInstance;
    }

    private WeakReference<Context> mContext;

    private PermissionUtils(Context ctx) {
        mContext = new WeakReference<Context> (ctx);
    }

    public boolean isPermissionGranted(String permission) {
        Context ctx = mContext.get();
        if (ctx == null) {
            return false;
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            int permissionStatus = ctx.checkSelfPermission(permission);
            Log.d(TAG, "permission:$permission result: " + permissionStatus);
            return permissionStatus == PackageManager.PERMISSION_GRANTED;
        } else {
            //permission is always granted at installation time on device <23
            return true;
        }
    }

    public List<String> getMissedPermissions() {
        List<String> missedPermissions = new ArrayList<> ();
        for (String permission : sRequiredPermissions) {
            if (!isPermissionGranted(permission)) {
                missedPermissions.add(permission);
            }
        }
        return missedPermissions;
    }
}

File: CameraTestApp/app/src/main/java/com/hank/cameratestapp/utils/PermissionUtils.kt
package com.hank.cameratestapp.utils

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log

interface PermissionUtils  {

    companion object {
        val TAG = PermissionUtils::class.java.simpleName

        private var instance: PermissionUtils? = null

        fun getInstance(context: Context): PermissionUtils {
            if (instance == null) {
                instance = PermissionUtilsImpl(context)
            }
            return instance as PermissionUtils
        }
    }

    fun getRequiredPermissions(): Array<String>
    fun isPermissionGranted(permission: String): Boolean
    /**
     * @return list of missing permissions
     */
    fun checkPermissions(requiredPermissions: Array<String>): List<String>

    private class PermissionUtilsImpl(private val context: Context): PermissionUtils {
        override fun getRequiredPermissions(): Array<String> {
            return arrayOf(Manifest.permission.CAMERA,
                          Manifest.permission.RECORD_AUDIO,
                          Manifest.permission.WRITE_EXTERNAL_STORAGE,
                          Manifest.permission.ACCESS_FINE_LOCATION)
        }

        override fun isPermissionGranted(permission: String): Boolean {
            return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                val result = context.checkSelfPermission(permission)
                Log.d(TAG, "permission:$permission result:$result")
                result == PackageManager.PERMISSION_GRANTED
            } else {
                //permission is always granted at installation time on device <23
                true
            }
        }

        override fun checkPermissions(requiredPermissions: Array<String>): List<String> {
            val missedPermissions = ArrayList<String>()

            for (permission in requiredPermissions) {
                if (!isPermissionGranted(permission)) {
                    missedPermissions.add(permission)
                }
            }

            return missedPermissions
        }
    }

}